bundles/org.openhab.core.io.rest.sse/src/main/java/org/openhab/core/io/rest/sse/SseStatesResource.java

```diff
@@ -0,0 +1,214 @@
+    public Object getStateEvents() throws IOException, InterruptedException {
+
+        final SseStateEventOutput eventOutput = new SseStateEventOutput();
+        broadcaster.add(eventOutput);
+
+        // Disables proxy buffering when using an nginx http server proxy for this response.
+        // This allows you to not disable proxy buffering in nginx and still have working sse
+        response.addHeader(X_ACCEL_BUFFERING_HEADER, "no");
+
+        // We want to make sure that the response is not compressed and buffered so that the client receives server sent
+        // events at the moment of sending them.
+        response.addHeader(HttpHeaders.CONTENT_ENCODING, "identity");
+
+        return eventOutput;
+    }
+
+    /**
+     * Alters the list of tracked items for a given state update connection
+     *
+     * @param connectionId the connection Id to change
+     * @param itemNames the list of items to track
+     */
+    @POST
+    @Path("{connectionId}")
+    @ApiOperation(value = "Changes the list of items a SSE connection will receive state updates to.")
+    @ApiResponses(value = { @ApiResponse(code = 200, message = "OK"),
+            @ApiResponse(code = 404, message = "Unknown connectionId") })
+    public Object updateTrackedItems(@PathParam("connectionId") String connectionId,
+            @ApiParam("items") Collection<String> itemNames) {
+
+        // Send the current states of the new tracked items on the SSE connection
+        Map<String, String> newTrackedItems = new HashMap<>();
+        itemNames.stream().forEach(i -> {
+            Item item = itemRegistry.get(i);
+            if (item != null) {
+                String transformedState = considerTransformation(item.getState().toString(), item, null);
+                newTrackedItems.put(item.getName(), transformedState);
+            }
+        });
+
+        try {
+            broadcaster.updateTrackedItems(connectionId, newTrackedItems);
+        } catch (IllegalArgumentException e) {
+            return Response.status(Status.NOT_FOUND).build();
+        }
+
+        return Response.ok().build();
+    }
+
+    /**
+     * Broadcasts a state event to all currently listening clients, after transforming it to a simple map.
+     *
+     * @param stateChangeEvent the {@link ItemStateChangedEvent} containing the new state
+     */
+    public void broadcastEvent(final ItemStateChangedEvent stateChangeEvent) {
+        executorService.execute(new Runnable() {
+            @Override
+            public void run() {
+                OutboundEvent.Builder builder = new OutboundEvent.Builder();
+                Map<String, String> payload = new HashMap<>();
+                Item item = itemRegistry.get(stateChangeEvent.getItemName());
+                if (item != null) {
+                    String transformedState = considerTransformation(item.getState().toString(), item, null);
+                    payload.put(item.getName(), transformedState);
```

Comment:
As mentioned in the issue, I am pretty sure that you will additionally require the raw state as well in the UI, e.g. for being able to request the correct item for the state, apply ui-based logic on visibility or colors, etc. So the current item->state map might be too simplistic after all.
Also, we might add further metadata to the `ItemStateChangedEvent` in future, such as the reason why it changed (manual command, rule triggered, or whatever). Serializing the whole content of ItemStateChangedEvent would somehow feel more future-proof, wdyt?

Reply:
> I am pretty sure that you will additionally require the raw state as well in the UI

True.

> Serializing the whole content of ItemStateChangedEvent would somehow feel more future-proof

The idea with this is not actually to expose events from the event bus, but only to make it as easy as possible for clients to keep track of the state of a predetermined set of items. So: 1) you open the SSE connection, 2) you send the list of items whose state you want to track, 3) you receive the current states of all your tracking list over the SSE connection, 4) you also receive any individual updates until you close the connection or change the tracking list.

The benefit with item-state maps is that for 3) ___and___ 4) the client-side code is pretty simple and actually the same in both cases: basically, just merge the map you received with the one you currently have. 3) is also useful because with this feature you might not even need to call `/rest/items` to initialize your client-side map if all you need are the current states. However, 3) doesn't correspond to any event on the event bus. That's why I wanted to abstract the event bus away from this API. Maybe the state could be an 2-element array or simple object (`{"state": "CLOSED", "transformedState": "zu"}`) in case there's a transformation involved.

Moreover it was asked a few years ago (IIRC by me) over in ESH to include the transformed state in `ItemStateChangedEvent`s but it wasn't accepted, because the transformation is possibly costly to compute (e.g. exec or JS) and blocking so it's not a good idea to do it on every change if it's not useful. Therefore, there's indeed a problem with this code here - it should first determine whether the item is being tracked or not.

Reply:
If we'd drop transformation support and thus would only have patterns, which you can render on client side, we could merely pass the raw state in your item state map. Would that solve all our issues?

Reply:
> If we'd drop transformation support and thus would only have patterns, which you can render on client side, we could merely pass the raw state in your item state map. Would that solve all our issues?

Turns out, even formatting the state using a pattern is no trivial problem (dates, quantity types unit conversions...), and it requires the client to know about the state description, which might not be always the case... so I figured making the server do the heavy lifting was the best option after all, and introduced in https://github.com/openhab/openhab-core/pull/1364/commits/1d11fa4e8076410e9dde8e3eb3243b944c041d09 a "displayState" in the map value.
So with items like this:
```
DateTime DemoDateTime { stateDescription="N/A" [ pattern="%1$tA, %1$td.%1$tm.%1$tY" ] }
Number:Temperature DemoTemperature "Info [%.5f %unit%]"
Number DemoFormat "Info [%.5f kWh]"
Contact DemoOptions "Options" { stateDescription="N/A" [ options="OPEN=unlocked,CLOSED=locked" ] }
```
It gives (line breaks added for clarity):
```
data: {
"DemoFormat":{"state":"500","displayState":"500.00000 kWh"},
"DemoContact":{"state":"OPEN"},
"DemoOptions":{"state":"CLOSED","displayState":"locked"},
"MetadataString":{"state":"1","displayState":"HDMI1"},
"DemoTemperature":{"state":"23.00 °F","displayState":"-5.00000 °C"},
"DemoDateTime":{"state":"2020-02-18T11:37:29.782672+0100","displayState":"Tuesday, 18.02.2020"}
}
```
So the UI can happily use the displayState for display purposes if it exists (or fall back to the actual state) and also have the raw version at hand.

Reply:
Sounds sensible then, thanks.